src/content/docs/en/reference/adapter-reference.mdx

```diff
@@ -188,3 +202,123 @@ You can usually call `app.render(request)` without using `.match` because Astro
 ```
 
 Once you [publish your adapter to npm](https://docs.npmjs.com/cli/v8/commands/npm-publish), running `astro add example` will install your package with any peer dependencies specified in your `package.json`. We will also instruct users to update their project config manually.
+
+## Adapter features
+
+A set of features that changes the output of the emitted files. When an adapter opts in to these features, they will get additional information inside specific hooks.
+
+### `functionPerRoute`
+
+This is a feature that is enabled when using SSR only. By default, Astro emits a single `entry.mjs` file, which is responsible to emit the rendered page on each request.
+
+When `functionPerRoute` is `true`, Astro will create a file for each route defined in the project. 
+Each file emitted will only render one page. The pages will be emitted inside a `dist/pages/` directory, and the emitted files will keep the same file paths of the `src/pages` directory.
+
+Enable the feature by passing `true` to the adapter.
+
+```js title="my-adapter.mjs" ins={9-11}
+export default function createIntegration() {
+  return {
+    name: '@matthewp/my-adapter',
+    hooks: {
+      'astro:config:done': ({ setAdapter }) => {
+        setAdapter({
+          name: '@matthewp/my-adapter',
+          serverEntrypoint: '@matthewp/my-adapter/server.js',
+          adapterFeatures: {
+              functionPerRoute: true
+          } 
+        });
+      },
+    },
+  };
+}
+```
+
+Then, consume the hook [`astro:build:ssr`](/en/reference/integrations-reference/#astrobuildssr), which will give you an `entryPoints`, an object that maps a page route to the physical file emitted after the build.
+
+```js title="my-adapter.mjs" ins={15-19}
+export default function createIntegration() {
+  return {
+    name: '@matthewp/my-adapter',
+    hooks: {
+      'astro:config:done': ({ setAdapter }) => {
+        setAdapter({
+          name: '@matthewp/my-adapter',
+          serverEntrypoint: '@matthewp/my-adapter/server.js',
+          adapterFeatures: {
+              functionPerRoute: true
+          } 
+        });
+      },
+
+      'astro:build:ssr': ({ entryPoints }) => {
+          for (const [route, entryFile] of entryPoints) {
+              // do something with route and entryFile
+          }
+      }  
+    },
+  };
+}
+```
+
+:::caution
+The `entryFile` is of type `URL` and represents the physical path of the file in the file system. This means that the paths change based on the OS where the code runs. 
+:::
+
+### `edgeMiddleware`
+
+Defines whether any SSR middleware code will be bundled when built.
+	
+When enabled, middleware code is not bundled and imported by all pages during the build:
```

Comment:
I just need clarification on this sentence @ematipico - is this: 

> "not bundled" ... and (but yes!) imported by all pages

> NOT "bundled and imported..." 

Why can't we just write logical arguments in docs :laughing:

Reply:
To be more clear: "not bundled and not imported." :D